File: Easy/563-Binary Tree Tilt/563.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    
    def findTilt(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        tilt, _ = self.DFS(root)
        return tilt
    
    def DFS(self, root):
        
        if not root:
            return [0, 0]
        
        ltilt, left  = self.DFS(root.left)
        rtilt, right = self.DFS(root.right)
        tilt = abs(left - right) + ltilt + rtilt
        
        return tilt, left + right + root.val

File: Easy/563-Binary Tree Tilt/563.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int findTilt(TreeNode* root) {
        
        if(!root)
            return 0;
        
        int tilt = findTilt(root->left) + findTilt(root->right);
        int lval = root->left  ? root->left->val  : 0;
        int rval = root->right ? root->right->val : 0;
        root->val += lval + rval;
        return abs(lval - rval) + tilt;
    }
};